File: admin/server/src/controllers/credit.controller.js
const Settings = require('../models/settings.model');
const CreditLedger = require('../models/creditLedger.model');
const CreditRequest = require('../models/creditRequest.model');
const User = require('../models/user.model');
const credits = require('../utils/credits');
const { ok, fail, asyncHandler } = require('../utils/response');

/**
 * Admin-only try-on credit management. Costing is tracked internally (BrandShoot
 * exposes no balance API): Purchased comes from Settings, Consumed is summed from
 * the ledger, Remaining = Purchased − Consumed.
 */

// GET /api/admin/credits/summary
const costingSummary = asyncHandler(async (_req, res) => {
  const s = (await Settings.findOne().lean()) || {};
  const purchased = Number(s.purchasedCredits || 0);
  const costPerCredit = Number(s.costPerCredit || 0);
  const currency = s.currency || 'INR';
  const consumed = await credits.getConsumedCredits();

  // If BrandShoot ever reports a live figure, trust it over the manually
  // entered total — that is the closest thing to a real sync available.
  const reported = s.brandshootReportedCredits;
  const hasReported = typeof reported === 'number' && Number.isFinite(reported);
  const remaining = hasReported ? reported : purchased - consumed;

  // User-side aggregates so the dashboard reflects real distribution, not just
  // the pool: how many credits are sitting in shopper balances right now, how
  // many were ever handed out, and the average per user.
  const agg = await User.aggregate([
    {
      $group: {
        _id: null,
        totalUsers: { $sum: 1 },
        allocated: { $sum: { $ifNull: ['$tryonCredits', 0] } },
        grantedTotal: { $sum: { $ifNull: ['$creditsGrantedTotal', 0] } },
        usedTotal: { $sum: { $ifNull: ['$creditsConsumedTotal', 0] } },
      },
    },
  ]);
  const a = agg[0] || { totalUsers: 0, allocated: 0, grantedTotal: 0, usedTotal: 0 };
  const avgPerUser = a.totalUsers ? a.allocated / a.totalUsers : 0;

  return ok(
    res,
    {
      purchased,
      consumed,
      remaining,
      costPerCredit,
      currency,
      estConsumedCost: consumed * costPerCredit,
      estRemainingValue: remaining * costPerCredit,
      // distribution
      totalUsers: a.totalUsers,
      allocated: a.allocated,
      grantedTotal: a.grantedTotal,
      usedTotal: a.usedTotal,
      avgPerUser,
      estAllocatedCost: a.allocated * costPerCredit,
      // live BrandShoot key status (auto-synced from real API calls)
      brandshootConfigured: !!s.brandshootKey,
      brandshootReportedCredits: hasReported ? reported : null,
      brandshootDepleted: !!s.brandshootDepleted,
      brandshootCheckedAt: s.brandshootCheckedAt || null,
      remainingIsLive: hasReported,
    },
    'Credit costing'
  );
});

// GET /api/admin/credits/users/:id/ledger — a user's credit history (newest first)
const userLedger = asyncHandler(async (req, res) => {
  const rows = await CreditLedger.find({ userId: req.params.id })
    .sort({ createdAt: -1 })
    .limit(200)
    .lean();
  return ok(res, rows, 'Credit history');
});

// POST /api/admin/credits/users/:id/adjust  body: { op: 'set'|'add'|'deduct', amount, reason }
const adjustUser = asyncHandler(async (req, res) => {
  const { op = 'add', amount, reason = '' } = req.body || {};
  const n = Number(amount);
  if (!Number.isFinite(n) || n < 0) {
    return fail(res, 'A non-negative amount is required.', 400);
  }
  if (!['set', 'add', 'deduct'].includes(op)) {
    return fail(res, "op must be 'set', 'add' or 'deduct'.", 400);
  }
  const result = await credits.adminAdjust({
    userId: req.params.id,
    op,
    amount: n,
    reason,
    createdBy: req.user.id,
  });
  if (!result) return fail(res, 'User not found.', 404);
  return ok(res, { tryonCredits: result.balanceAfter, entry: result.ledger }, 'Credits updated');
});

// POST /api/admin/credits/purchase  body: { amount, reason } — record a pool top-up
const purchase = asyncHandler(async (req, res) => {
  const n = Number((req.body || {}).amount);
  if (!Number.isFinite(n) || n <= 0) {
    return fail(res, 'A positive amount is required.', 400);
  }
  const purchasedCredits = await credits.recordPoolPurchase({
    amount: n,
    reason: (req.body || {}).reason || '',
    createdBy: req.user.id,
  });
  return ok(res, { purchasedCredits }, 'Purchase recorded');
});

// POST /api/admin/credits/pool  body: { purchasedCredits?, costPerCredit?, currency? }
// Set the pool to the credits the BrandShoot key actually carries (e.g. 10000)
// and the price per credit. Used by the Credit pool card on the Credits page.
const setPool = asyncHandler(async (req, res) => {
  const b = req.body || {};
  const update = {};
  if (b.purchasedCredits != null) {
    const n = Number(b.purchasedCredits);
    if (!Number.isFinite(n) || n < 0) {
      return fail(res, 'purchasedCredits must be 0 or more.', 400);
    }
    update.purchasedCredits = Math.round(n);
  }
  if (b.costPerCredit != null) {
    const c = Number(b.costPerCredit);
    if (!Number.isFinite(c) || c < 0) {
      return fail(res, 'costPerCredit must be 0 or more.', 400);
    }
    update.costPerCredit = c;
  }
  if (b.currency != null) update.currency = String(b.currency).trim() || 'INR';
  if (!Object.keys(update).length) return fail(res, 'Nothing to update.', 400);

  const doc = await Settings.findOneAndUpdate({}, { $set: update }, { new: true, upsert: true });
  return ok(
    res,
    {
      purchasedCredits: doc.purchasedCredits,
      costPerCredit: doc.costPerCredit,
      currency: doc.currency,
    },
    'Credit pool updated'
  );
});

// GET /api/admin/credits/users?search= — every user with their credit balance,
// for the admin "distribute credits" section.
const listUsers = asyncHandler(async (req, res) => {
  const search = String(req.query.search || '').trim();
  const filter = search
    ? {
        $or: [
          { Email: { $regex: search, $options: 'i' } },
          { FirstName: { $regex: search, $options: 'i' } },
          { LastName: { $regex: search, $options: 'i' } },
        ],
      }
    : {};
  const limit = Math.min(Math.max(parseInt(req.query.limit, 10) || 200, 1), 500);
  const rows = await User.find(filter)
    .select('FirstName LastName Email ProfilePicture Role tryonCredits creditsGrantedTotal creditsConsumedTotal')
    .sort({ createdAt: -1 })
    .limit(limit)
    .lean();
  return ok(res, rows, 'Users with credits');
});

// GET /api/admin/credits/requests?status=pending|approved|rejected|all
const listRequests = asyncHandler(async (req, res) => {
  const status = String(req.query.status || 'all');
  const filter = ['pending', 'approved', 'rejected'].includes(status) ? { status } : {};
  const items = await CreditRequest.find(filter).sort({ createdAt: -1 }).limit(200).lean();
  const pendingCount = await CreditRequest.countDocuments({ status: 'pending' });
  return ok(res, { items, pendingCount }, 'Credit requests');
});

// POST /api/admin/credits/requests/:id  body: { action:'approve'|'reject', amount?, note? }
const handleRequest = asyncHandler(async (req, res) => {
  const { action, amount, note = '' } = req.body || {};
  if (!['approve', 'reject'].includes(action)) {
    return fail(res, "action must be 'approve' or 'reject'.", 400);
  }
  const doc = await CreditRequest.findById(req.params.id);
  if (!doc) return fail(res, 'Request not found.', 404);
  if (doc.status !== 'pending') return fail(res, 'This request was already handled.', 409);

  let balanceAfter = null;
  if (action === 'approve') {
    const give = Math.max(1, Math.round(Number(amount) || doc.amount));
    const result = await credits.adminAdjust({
      userId: doc.userId,
      op: 'add',
      amount: give,
      reason: note || 'credit request approved',
      createdBy: req.user.id,
    });
    if (!result) return fail(res, 'User not found.', 404);
    balanceAfter = result.balanceAfter;
    doc.grantedAmount = give;
  }
  doc.status = action === 'approve' ? 'approved' : 'rejected';
  doc.adminNote = String(note || '').slice(0, 500);
  doc.handledBy = req.user.id;
  doc.handledAt = new Date();
  await doc.save();

  return ok(
    res,
    { request: doc, balanceAfter },
    action === 'approve' ? 'Credits granted' : 'Request rejected'
  );
});

// DELETE /api/admin/credits/requests/:id
const deleteRequest = asyncHandler(async (req, res) => {
  const doc = await CreditRequest.findByIdAndDelete(req.params.id);
  if (!doc) return fail(res, 'Request not found.', 404);
  return ok(res, { id: req.params.id }, 'Request deleted');
});

// POST /api/admin/credits/grant-all  body: { amount, onlyUngranted? }
// Bulk-distribute credits: to everyone, or only to users who never got any.
const grantAll = asyncHandler(async (req, res) => {
  const b = req.body || {};
  const n = Math.round(Number(b.amount) || 0);
  if (!(n > 0)) return fail(res, 'A positive amount is required.', 400);
  const fn = b.onlyUngranted ? credits.grantToAllUngranted : credits.grantToEveryone;
  const granted = await fn({
    amount: n,
    reason: b.reason || 'admin bulk distribution',
    createdBy: req.user.id,
  });
  return ok(res, { granted, amount: n }, `Granted ${n} credit(s) to ${granted} user(s)`);
});

module.exports = {
  costingSummary,
  userLedger,
  adjustUser,
  purchase,
  setPool,
  listUsers,
  listRequests,
  handleRequest,
  deleteRequest,
  grantAll,
};

File: admin/lib/features/personalization/controllers/credit_controller.dart
import 'package:get/get.dart';

import '../../../data/repositories/credits/credit_ledger_model.dart';
import '../../../data/repositories/credits/credit_repository.dart';
import '../../../data/repositories/credits/credit_request_model.dart';
import '../../../data/repositories/credits/credit_user_model.dart';
import '../../../utils/popups/loaders.dart';

/// Manages the try-on credit costing dashboard (pool-level) and per-user credit
/// history/adjustments shown on the customer detail screen.
class CreditController extends GetxController {
  static CreditController get instance => Get.find();

  final CreditRepository _repo = Get.put(CreditRepository());

  // ---- Costing summary (pool) ----
  final RxBool summaryLoading = false.obs;
  final RxInt purchased = 0.obs;
  final RxInt consumed = 0.obs;
  final RxInt remaining = 0.obs;
  final RxDouble costPerCredit = 0.0.obs;
  final RxString currency = 'INR'.obs;
  final RxDouble estConsumedCost = 0.0.obs;
  final RxDouble estRemainingValue = 0.0.obs;
  // Distribution aggregates
  final RxInt totalUsers = 0.obs;
  final RxInt allocated = 0.obs;
  final RxInt grantedTotal = 0.obs;
  final RxInt usedTotal = 0.obs;
  final RxDouble avgPerUser = 0.0.obs;
  final RxDouble estAllocatedCost = 0.0.obs;
  final RxBool savingPool = false.obs;

  // ---- Per-user (customer detail) ----
  final RxBool ledgerLoading = false.obs;
  final RxList<CreditLedgerModel> ledger = <CreditLedgerModel>[].obs;
  final RxInt userBalance = 0.obs;
  final RxBool adjusting = false.obs;

  // ---- Distribution table (all users) ----
  final RxBool usersLoading = false.obs;
  final RxList<CreditUserModel> users = <CreditUserModel>[].obs;
  final RxString userSearch = ''.obs;

  // ---- Shopper credit requests (admin inbox) ----
  final RxBool requestsLoading = false.obs;
  final RxList<CreditRequestModel> requests = <CreditRequestModel>[].obs;
  final RxInt pendingCount = 0.obs;
  final RxString requestFilter = 'all'.obs;

  // ---- Try-on model curation ----
  final RxBool modelsLoading = false.obs;
  final RxList<Map<String, dynamic>> brandshootModels = <Map<String, dynamic>>[].obs;
  final RxSet<String> enabledModelIds = <String>{}.obs;
  final RxBool savingModels = false.obs;

  @override
  void onInit() {
    fetchSummary();
    super.onInit();
  }

  /// Load everything the Credits screen needs.
  Future<void> loadAll() async {
    await Future.wait([
      fetchSummary(),
      fetchUsers(),
      fetchRequests(),
      fetchModels(),
    ]);
  }

  /// Save the pool total / price per credit, then refresh the dashboard.
  Future<void> savePool({
    int? purchasedCredits,
    double? costPerCredit,
    String? currency,
  }) async {
    try {
      savingPool.value = true;
      await _repo.setPool(
        purchasedCredits: purchasedCredits,
        costPerCredit: costPerCredit,
        currency: currency,
      );
      await fetchSummary();
      TLoaders.successSnackBar(
          title: 'Pool updated', message: 'Credit pool and pricing saved.');
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    } finally {
      savingPool.value = false;
    }
  }

  // ---- Users -------------------------------------------------------------

  Future<void> fetchUsers() async {
    try {
      usersLoading.value = true;
      users.value = await _repo.getUsers(search: userSearch.value);
    } catch (e) {
      users.clear();
    } finally {
      usersLoading.value = false;
    }
  }

  /// Add / set / deduct a user's credits straight from the distribution table.
  Future<void> adjustUser(String userId,
      {required String op, required int amount, String reason = ''}) async {
    try {
      adjusting.value = true;
      final newBalance = await _repo.adjustUserCredits(userId,
          op: op, amount: amount, reason: reason);
      final i = users.indexWhere((u) => u.id == userId);
      if (i != -1) {
        users[i].tryonCredits = newBalance;
        users.refresh();
      }
      await fetchSummary();
      TLoaders.successSnackBar(
          title: 'Credits updated', message: 'Balance is now $newBalance.');
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    } finally {
      adjusting.value = false;
    }
  }

  /// Bulk-distribute credits to everyone (or only to users who never got any).
  Future<void> grantAll(int amount, {bool onlyUngranted = false}) async {
    try {
      adjusting.value = true;
      final n = await _repo.grantAll(amount: amount, onlyUngranted: onlyUngranted);
      await fetchUsers();
      await fetchSummary();
      TLoaders.successSnackBar(
          title: 'Credits distributed',
          message: 'Gave $amount credit(s) to $n user(s).');
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    } finally {
      adjusting.value = false;
    }
  }

  // ---- Requests ----------------------------------------------------------

  Future<void> fetchRequests() async {
    try {
      requestsLoading.value = true;
      final r = await _repo.getRequests(status: requestFilter.value);
      requests.value = r.items;
      pendingCount.value = r.pendingCount;
    } catch (e) {
      requests.clear();
    } finally {
      requestsLoading.value = false;
    }
  }

  Future<void> handleRequest(String id,
      {required String action, int? amount, String note = ''}) async {
    try {
      await _repo.handleRequest(id, action: action, amount: amount, note: note);
      await fetchRequests();
      await fetchUsers();
      await fetchSummary();
      TLoaders.successSnackBar(
        title: action == 'approve' ? 'Approved' : 'Rejected',
        message: action == 'approve'
            ? 'Credits granted to the shopper.'
            : 'The request was rejected.',
      );
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    }
  }

  Future<void> deleteRequest(String id) async {
    try {
      await _repo.deleteRequest(id);
      await fetchRequests();
      TLoaders.successSnackBar(title: 'Deleted', message: 'Request removed.');
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    }
  }

  // ---- Model curation ----------------------------------------------------

  Future<void> fetchModels() async {
    try {
      modelsLoading.value = true;
      final results = await Future.wait([
        _repo.getBrandshootModels(),
        _repo.getEnabledModelIds(),
      ]);
      brandshootModels.value = results[0] as List<Map<String, dynamic>>;
      enabledModelIds.assignAll(results[1] as List<String>);
    } catch (e) {
      brandshootModels.clear();
    } finally {
      modelsLoading.value = false;
    }
  }

  void toggleModel(String id) {
    if (enabledModelIds.contains(id)) {
      enabledModelIds.remove(id);
    } else {
      enabledModelIds.add(id);
    }
  }

  Future<void> saveModels() async {
    try {
      savingModels.value = true;
      await _repo.saveEnabledModelIds(enabledModelIds.toList());
      TLoaders.successSnackBar(
        title: 'Saved',
        message: enabledModelIds.isEmpty
            ? 'No models selected — shoppers see the default women models.'
            : '${enabledModelIds.length} model(s) published to try-on.',
      );
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    } finally {
      savingModels.value = false;
    }
  }

  int _int(dynamic v) => (num.tryParse((v ?? 0).toString()) ?? 0).toInt();
  double _double(dynamic v) => (num.tryParse((v ?? 0).toString()) ?? 0).toDouble();

  Future<void> fetchSummary() async {
    try {
      summaryLoading.value = true;
      final d = await _repo.getCostingSummary();
      purchased.value = _int(d['purchased']);
      consumed.value = _int(d['consumed']);
      remaining.value = _int(d['remaining']);
      costPerCredit.value = _double(d['costPerCredit']);
      currency.value = (d['currency'] ?? 'INR').toString();
      estConsumedCost.value = _double(d['estConsumedCost']);
      estRemainingValue.value = _double(d['estRemainingValue']);
      totalUsers.value = _int(d['totalUsers']);
      allocated.value = _int(d['allocated']);
      grantedTotal.value = _int(d['grantedTotal']);
      usedTotal.value = _int(d['usedTotal']);
      avgPerUser.value = _double(d['avgPerUser']);
      estAllocatedCost.value = _double(d['estAllocatedCost']);
    } catch (_) {
      // non-fatal — dashboard just shows last-known / zeros
    } finally {
      summaryLoading.value = false;
    }
  }

  Future<void> loadUserLedger(String userId, {int? currentBalance}) async {
    try {
      ledgerLoading.value = true;
      if (currentBalance != null) userBalance.value = currentBalance;
      ledger.value = await _repo.getUserLedger(userId);
    } catch (_) {
      ledger.clear();
    } finally {
      ledgerLoading.value = false;
    }
  }

  /// Adjust a user's balance then refresh the ledger + costing summary.
  Future<void> adjust(
    String userId, {
    required String op,
    required int amount,
    String reason = '',
  }) async {
    try {
      adjusting.value = true;
      final newBalance =
          await _repo.adjustUserCredits(userId, op: op, amount: amount, reason: reason);
      userBalance.value = newBalance;
      await loadUserLedger(userId);
      await fetchSummary();
      TLoaders.successSnackBar(
          title: 'Credits updated', message: 'Balance is now $newBalance.');
    } catch (e) {
      TLoaders.errorSnackBar(title: 'Oh Snap', message: e.toString());
    } finally {
      adjusting.value = false;
    }
  }
}
